File: 0282-expression-add-operators/0282-expression-add-operators.cpp
class Solution {
public:
    vector<string> addOperators(string num, int target) {
        vector<string> result;
        backtrack(result, "", num, target, 0, 0, 0);
        return result;
    }

private:
    void backtrack(vector<string>& result, string expr, string num, int target, int pos, long currValue, long lastOperand) {
        if (pos == num.size()) {
            if (currValue == target) {
                result.push_back(expr);
            }
            return;
        }

        for (int i = pos; i < num.size(); ++i) {
            // Avoid numbers with leading zeros
            if (i != pos && num[pos] == '0') break;

            string currStr = num.substr(pos, i - pos + 1);
            long currNum = stol(currStr);

            if (pos == 0) {
                // First number, no operator
                backtrack(result, currStr, num, target, i + 1, currNum, currNum);
            } else {
                // Addition
                backtrack(result, expr + "+" + currStr, num, target, i + 1, currValue + currNum, currNum);
                // Subtraction
                backtrack(result, expr + "-" + currStr, num, target, i + 1, currValue - currNum, -currNum);
                // Multiplication
                backtrack(result, expr + "*" + currStr, num, target, i + 1, currValue - lastOperand + lastOperand * currNum, lastOperand * currNum);
            }
        }
    }
};

File: 0282-expression-add-operators/0282-expression-add-operators.py
from typing import List

class Solution:
    def addOperators(self, num: str, target: int) -> List[str]:
        result=[]

        def backtrack(index: int,path: str, value: int,last_operand: int):
            if index ==len(num):
                if value==target:
                    result.append(path)
                return
            for i in range(index,len(num)):
                if i != index and num[index]=='0':
                    break
                curr_str = num[index:i+1]
                curr_num = int(curr_str)

                if index ==0:
                    backtrack(i+1,curr_str,curr_num,curr_num)
                else:
                    backtrack(i + 1, path + '+' + curr_str, value + curr_num, curr_num)
                    
                    backtrack(i + 1, path + '-' + curr_str, value - curr_num, -curr_num)
                    
                    backtrack(i + 1, path + '*' + curr_str, value - last_operand + last_operand * curr_num, last_operand * curr_num)

        backtrack(0, "", 0, 0)
        return result

                

        